packages/components/src/icon/index.js

```diff
@@ -14,13 +14,39 @@ import { SVG } from '@wordpress/primitives';
  */
 import Dashicon from '../dashicon';
 
+/* eslint-disable jsdoc/valid-types */
+/**
+ * @template T
+ * @typedef {T extends import('react').ComponentType<infer U> ? U : T extends string ? import('react').ComponentPropsWithoutRef<'span'> : {}} AdditionalProps
```

Comment:
Should the final default be `never` rather than `{}`? When would we end up in that final branch?

Also an aside, the formatting like this is so frustrating to read... I'm assuming there's no way to split this across multiple lines with JSDoc right? Is that another limitation we could add to the list of reasons why full TS is preferred?

Reply:
> Should the final default be `never` rather than `{}`? When would we end up in that final branch?

🤔  This probably isn't perfect, I don't think it captures the case where icon is an element but this is a good start.

We want something that we can intersect (`&`) with another type to just keep the other type (an identity element). That's `{}` (I don't know whether there are other types that would work) where `A & {}` is the same as `A`. `A & never` is `never` (I had to [play a bit to test this out](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UDeUCGAuKByVWoF8AoUSKAIQWSgCNMtq8iToYLE4AycgbmPBdSUYAbRxYAulAD0UqACJUcvqTKDEZUQ0kz51Jc1iqhGsdtkL9-Q9WOaJ083uUsAdhABuEAE5CoXN55e3A5QAd7OsBAAtmCgvlxI+CEwhEA)).

> Also an aside, the formatting like this is so frustrating to read... I'm assuming there's no way to split this across multiple lines with JSDoc right? Is that another limitation we could add to the list of reasons why full TS is preferred?

💯  It's very unpleasant. I don't believe TypeScript + JSDoc work properly when we break these on multiple lines. It would be much nicer in TypeScript. For reference, this is equivalent TypeScript:

```ts
export type AdditionalProps<T> =
    T extends React.ComponentType<infer U>
    ? U
    : T extends string
    ? React.ComponentPropsWithoutRef<'span'>
    : {};
```

Fortunately, although it's exported (JSDoc typedefs are always exported, but this likely isn't easily accessible from the package) it's really an internal type helper.